### ISSUE TITLE
[BUG] Broken CosmosDB Spring Data Repository entity projection

### ISSUE THREAD

**Issue by PavelKhatsimtsau** [Azure/azure-sdk-for-java#20319]
Hi. We have encountered an issue related to inconsistent entity properties projection while switching to the latest version of azure-spring-data-cosmos library.
For example, given that CosmosDB spring data repository interface contains below method - 

```
@Query("SELECT VALUE c.productId FROM c WHERE c.uid = @uid")
String getProductIdByUID(@Param("id") String uid);
```

Once such a method will be executed the following exception will be raised:

```
java.lang.IllegalArgumentException: Entity is null.
    at org.springframework.util.Assert.notNull(Assert.java:201)
    at com.azure.spring.data.cosmos.core.convert.MappingCosmosConverter.read(MappingCosmosConverter.java:65)
    at com.azure.spring.data.cosmos.core.CosmosTemplate.toDomainObject(CosmosTemplate.java:893)
    at com.azure.spring.data.cosmos.core.CosmosTemplate.lambda$runQuery$23(CosmosTemplate.java:774)
```
	
It is caused by broken CosmosPersistentEntity mapping by 'Repository Entity'/'Repository derived method return' types.
Our assumption - the root cause of such a behavior was introduced during below code updates related to CosmosTemplate class:

https://github.com/Azure/azure-sdk-for-java/commit/3719570edb49ecceb8c8a874ae60a6c90fa9761a#diff-9d1b72616d654841a02e8a58ff64d4121a68b5009a9f62a9a233dae5df119403

Thanks!

**Comment by joshfree** [comment]
@kushagraThapar PTAL

**Comment by kushagraThapar** [comment]
@PavelKhatsimtsau - Is this issue happening consistently? What do you mean by `inconsistent entity properties projection` ?

cc - @Blackbaud-MikeLueders - Can you please help taking a look at this issue?
Is it a regression in this PR - https://github.com/Azure/azure-sdk-for-java/pull/19438 ?

**Comment by PavelKhatsimtsau** [comment]
Hi, everyone! Sorry for delayed respond.

I have tried to reproduce above problem with the latest 3.6.0 released version - and it is still an actual issue.
I am agree - inconsistent entity properties projection - is a bit frustrated term - let me explain what I exactly was trying to describe:

Let`s assume - that we have cosmos db repository for some abstract Product type with below methods:

```
@Query("SELECT VALUE c.productId FROM c WHERE c.uid = @productId")
String getProductIdByUID(@Param("productId") String uid);

@Query("SELECT VALUE COUNT(1) FROM c WHERE c.uid= @productId")
Long countByProductId(@Param("productId") String uid);

```

For both above methods MappingContext<? extends CosmosPersistentEntity<?>,  CosmosPersistentProperty> mappingContext field from com.azure.spring.data.cosmos.core.convert.MappingCosmosConverter will contain the following entity mapping items:

- 	{ClassTypeInformation@11111} "java.lang.String" -> {Optional@22222} "Optional.empty"
- 	{ClassTypeInformation@33333} "java.lang.Long" -> {Optional@44444} "Optional.empty"
- 	{ClassTypeInformation@18397} "com.example..model.Product" -> {Optional@55555} "Optional[com.azure.spring.data.cosmos.core.mapping.BasicCosmosPersistentEntity@5173ef95]"

So when methods have some property model projection or any aggregated value as its return type  - empty CosmosPersistentEntity<?> object will be returned in scope of below code

com.azure.spring.data.cosmos.core.convert.MappingCosmosConverter#read:64


```
@Override
    public <R> R read(Class<R> type, JsonNode jsonNode) {

        final CosmosPersistentEntity<?> entity = mappingContext.getPersistentEntity(type);
        Assert.notNull(entity, "Entity is null.");

        return readInternal(entity, type, jsonNode);
    }
```

Hopefully it will be useful information to reveal the problem we encountered. Thanks!

**Comment by kushagraThapar** [comment]
@PavelKhatsimtsau - thanks for the explanation, I understand the issue somewhat now. Would it be possible for you to provide a sample repo for this issue ? So that I can look into a possible fix for this.

**Comment by snimmagadda1** [comment]
circling back much later, has anyone gotten projections to work successfully in latest versions? 

I am running into similar issues as above.. It looks like the implementation is not respecting the return type of the projection, and is instead returning the entire entity

**Comment by kushagraThapar** [comment]
@trande4884 can you please take a look at this issue, thanks!

**Comment by trande4884** [comment]
@snimmagadda1 is there a sample repo you can link that shows this issue?

I ask because our integration tests have functions that are the same as what you are listing and they do not throw this error:
`@Query(value = "SELECT VALUE COUNT(1) from c where c.title = @title")
Long countByQueryWithNonPrimitive(@Param("title") String title);`

**Comment by PlayAnyData** [comment]
Hey,

I have recently stumbled over a similar issue as well, not sure if it is the one above though. I have created a sample repository here: https://github.com/PlayAnyData/demo-cosmos-spring

From my point of view, there are some things that are not that obvious in the first place. I have added three different queries there, from which the first one throws the Exception:

`@Query("SELECT count(1) FROM c")
 long countObjects();`
Compared to @trande4884 integration test, we are missing the `VALUE` in here, which is a keyword for returning the value instead of the json (see here https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/query/select) and was present in the original poster's issue.

The second query here, does not work as well, although we have the VALUE keyword in there. I guess the problem is with the alias (AS). However, be aware that in case you have an ALIAS, it seems the VALUE keyword with CosmosRepository does not work as expected.
`@Query("SELECT VALUE count(1) as counter FROM c")
long countObjects2();`

This third one works fine. Although this is not exactly elegant, as now we are returning a projection as a list. Afterwards, we need to access the first object from it to get the value. MyDataProjection in this case is just containing one object named counter.
`@Query("SELECT count(1) as counter FROM c")
 List<MyDataProjection> countObjectsWrapped();`

I hope this sheds some lights into what might have caused the issue?!
